controllers/nodefeaturediscovery_finalizers.go

```diff
@@ -0,0 +1,411 @@
+	// Attempt to delete all components. If it fails, return
+	// a warning letting users know the deletion failed, and
+	// then call the reconciler once more to see if the error
+	// can be corrected.
+	klog.Info("Attempting to delete NFD operator components")
+	if err := r.deleteComponents(ctx, instance); err != nil {
+		klog.Error(err, "Failed to delete one or more components")
+		return ctrl.Result{}, err
+	}
+
+	// Check if all components are deleted. If they're not,
+	// then call the reconciler but wait 10 seconds before
+	// checking again.
+	klog.Info("Deletion appears to have succeeded, but running a secondary check to ensure resources are cleaned up")
+	if r.doComponentsExist(ctx, instance) {
+		klog.Info("Some components still exist. Requeueing deletion request.")
+		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	// If all components are deleted, then remove the finalizer
+	klog.Info("Secondary check passed. Removing finalizer if it exists.")
+	if r.hasFinalizer(instance, finalizer) {
+		r.removeFinalizer(instance, finalizer)
+		if err := r.Update(ctx, instance); err != nil {
+			if k8serrors.IsNotFound(err) {
+				return ctrl.Result{Requeue: false}, nil
+			}
+			klog.Info("Finalizer was found, but removing it was unsuccessful. Requeueing deletion request.")
+			return ctrl.Result{}, nil
+		}
+
+		klog.Info("Finalizer was found and successfully removed.")
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	klog.Info("Finalizer does not exist, but resource deletion succesful.")
+	return ctrl.Result{Requeue: false}, nil
+}
+
+// addFinalizer adds a finalizer to the NFD Operator instance
+func (r *NodeFeatureDiscoveryReconciler) addFinalizer(ctx context.Context, instance *nfdv1.NodeFeatureDiscovery, finalizer string) (ctrl.Result, error) {
+	// Add the defined finalizer as a finalizer to the instance if it does not exist
+	instance.Finalizers = append(instance.Finalizers, finalizer)
+	instance.Status.Conditions = r.getProgressingConditions("DeploymentStarting", "Deployment is starting")
+	if err := r.Update(ctx, instance); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// we exit reconcile loop because we will have additional update reconcile
+	return ctrl.Result{Requeue: false}, nil
+}
+
+// hasFinalizer determines if the operator instance has a specific
+// finalizer value, which is defined by the parameter 'finalizer'
+func (r *NodeFeatureDiscoveryReconciler) hasFinalizer(instance *nfdv1.NodeFeatureDiscovery, finalizer string) bool {
+	if len(instance.Finalizers) == 0 {
+		return false
+	}
+
+	// The instance will have a list of finalizers under its
+	// `metav1.ObjectMeta` reference
+	for _, f := range instance.Finalizers {
+
+		// If the current finalizer in the list matches the
+		// 'finalizer' parameter, then the operator does have
+		// the desired finalizer, so return "true"
+		if f == finalizer {
+			return true
+		}
+	}
+
+	// Return false, as the finalizer was not found in the list.
+	return false
```

Comment:
I think checking the length is redundant here. Also drop the excessive code comments

```suggestion
	for _, f := range instance.Finalizers {
		if f == finalizer {
			return true
		}
	}
	return false
```